File: src/AgentBin.Go/echoagent.go
package main

import (
	"context"
	"fmt"
	"iter"

	"github.com/a2aproject/a2a-go/v2/a2a"
	"github.com/a2aproject/a2a-go/v2/a2asrv"
)

type echoAgent struct{}

var _ a2asrv.AgentExecutor = (*echoAgent)(nil)

func (e *echoAgent) Execute(_ context.Context, execCtx *a2asrv.ExecutorContext) iter.Seq2[a2a.Event, error] {
	text := extractText(execCtx.Message)
	return func(yield func(a2a.Event, error) bool) {
		reply := a2a.NewMessage(a2a.MessageRoleAgent,
			a2a.NewTextPart(fmt.Sprintf("Echo: %s", text)))
		yield(reply, nil)
	}
}

func (e *echoAgent) Cancel(_ context.Context, _ *a2asrv.ExecutorContext) iter.Seq2[a2a.Event, error] {
	return func(yield func(a2a.Event, error) bool) {}
}

File: src/AgentBin.Python/echo_agent.py
"""EchoAgent executor for AgentBin Python server."""

from a2a.server.agent_execution.agent_executor import AgentExecutor
from a2a.server.agent_execution.context import RequestContext
from a2a.server.events.event_queue import EventQueue
from a2a.types import Message, Part, Role


class EchoAgent(AgentExecutor):
    """Simple echo agent that returns the user's message."""

    async def execute(self, context: RequestContext, event_queue: EventQueue) -> None:
        """Echo back the user's message."""
        text = extract_text(context.message)
        
        reply = Message(
            role=Role.ROLE_AGENT,
            message_id=f"echo-{context.message.message_id}",
            task_id=context.task_id,
            context_id=context.context_id,
            parts=[Part(text=f"Echo: {text}")],
        )
        
        await event_queue.enqueue_event(reply)

    async def cancel(self, context: RequestContext, event_queue: EventQueue) -> None:
        """Cancel not implemented for echo agent."""
        pass


def extract_text(message: Message) -> str:
    """Extract the first text part from a message."""
    if not message or not message.parts:
        return ""
    
    for part in message.parts:
        if part.HasField("text"):
            return part.text
    
    return ""
